journey-builder-triggers/template.ts

```diff
@@ -0,0 +1,161 @@
+import { RecommendationsConfig, RecipeReferenceLookup, RecipeReference, recommend } from "recs";
+import { UserAttributeLookup, UserAttributeReference, UserSegmentLookup, UserSegmentReference } from "common";
+
+export class JourneyBuilderTriggersUtils {
+    static defaultCatalogObject = "Product";
+    static defaultCatalogObjectAttributes = ["id", "name", "imageUrl", "url", "price"];
+    static maxRecommendations = 8;
+
+    // ================================== Catalog Localization =====================================
+    // If you use catalog localization and would like to receive localized catalog data
+    // in Marketing Cloud, change the value of 'isCatalogLocalized' to true
+    // =============================================================================================
+    static isCatalogLocalized = false;
+
+    static getFlatItems(context: CampaignComponentContext, catalogObject: string, ids: string[]): any[] {
+        const locale: string = context.locale || "";
+
+        return context.services.catalog
+            .findItems(catalogObject, ids)
+            .map(item => item.toFlatJSON(JourneyBuilderTriggersUtils.defaultCatalogObjectAttributes, JourneyBuilderTriggersUtils.isCatalogLocalized ? locale : null));
+    }
+
+    static getUserAttributeValue(attributeId: string, user: User) {
+        return (user?.attributes?.[attributeId] as Attribute|undefined)?.value?.toString() || "";
+    }
+
+    static getUserSegments(segments: UserSegmentReference[], user: User) {
+        if (segments?.length && user?.segmentMembership?.length) {
+            const selectedSegments: string[] = segments.map(s => s.id);
+            return user.segmentMembership.filter(sm => selectedSegments.includes(sm.segmentId)).map(sm => sm.segmentName).join(",");
+        }
+        return "";
+    }
+
+    static getRecommendations(recipe: RecipeReference, context: CampaignComponentContext) {
+        let result = [];
+
+        if (recipe) {
+            const recipeConfig = new RecommendationsConfig();
+            recipeConfig.maxResults = JourneyBuilderTriggersUtils.maxRecommendations;
+            recipeConfig.recipe = recipe;
+
+            try {
+                if (recipeConfig?.recipe?.id) {
+                        const recs = recommend(context, recipeConfig).map(i => i.id);
+                        result = JourneyBuilderTriggersUtils.getFlatItems(context, recipeConfig.itemType, recs);
+                }
+
+                return JSON.stringify(result);
+
+            } catch(ex){}
+        }
+
+        return JSON.stringify(result);
```

Comment:
Why are we only catching one possible exception and not another?

It seems like we're not potentially catching the second use of `stringify` because `result` would be an empty array but I'd almost rather see this instead:

```suggestion
            try {
                if (recipeConfig?.recipe?.id) {
                        const recs = recommend(context, recipeConfig).map(i => i.id);
                        result = JourneyBuilderTriggersUtils.getFlatItems(context, recipeConfig.itemType, recs);
                }
            } catch (ex) {
                return JSON.stringify([]);
            }
        }

        return JSON.stringify(result);
```

Reply:
We should just be returning a default value in all cases where there is a potential exception/issue.